GitHub Actions Naming Scheme

PhilippSchmelter opened the pull request:
Solves #1010 
Merge #982 First!!!!!!

sebastian-peter commented on build.gradle:
```diff
@@ -54,6 +54,7 @@ apply from: scriptsLocation + 'scoverage.gradle' // scoverage scala code coverag
 apply from: scriptsLocation + 'deploy.gradle'
 apply from: scriptsLocation + 'semVer.gradle'
 apply from: scriptsLocation + 'mavenCentralPublish.gradle'
+apply from: scriptsLocation + 'BranchName.gradle'
```
Please start the file name with a lower case letter

sebastian-peter commented on gradle/scripts/BranchName.gradle:
```diff
@@ -0,0 +1,23 @@
+    println "Branch name is $branchName"
+    project.ext.set('branch_name', branchName)
+  }
+}
```
Last empty line ;)

sebastian-peter replied:
Still valid :)

sebastian-peter commented on gradle/scripts/BranchName.gradle:
```diff
@@ -0,0 +1,23 @@
+task checkBranchName {
```
Please try `tasks.register('checkBranchName')` as Intellij suggests

sebastian-peter commented on gradle/scripts/BranchName.gradle:
```diff
@@ -0,0 +1,23 @@
+      branchName = branchName.replaceFirst('refs/heads/', '')
+    }
+
+    def pattern = /^[a-z]{2}\/#([0-9]+)(-.*)?$/
```
The capture groups don't hurt much, but are also not necessary I think.

This might work too: `^[a-z]{2}\/#[0-9]+(?:-.+)?$`

sebastian-peter commented on gradle/scripts/BranchName.gradle:
```diff
@@ -0,0 +1,23 @@
+task checkBranchName {
+  doLast {
+    def branchName = System.getenv('GITHUB_REF_NAME')
+    def eventName = System.getenv('GITHUB_EVENT_NAME')
+    def headRef = System.getenv('GITHUB_HEAD_REF')
```
Would be great if you could pass the branch name variable to the task like I mentioned in [this comment](https://github.com/ie3-institute/simona/pull/1011#issuecomment-2483029978). (I did not point to the right resource though, it should be [this](https://docs.gradle.org/current/userguide/project_properties.html))

So, it'd be great to find the proper branch name within ci.yml, and then call the task somehow like this:
```bash
./gradlew -PbranchName=[...] checkBranchName
```

So this way, we would handle GitHub-specific stuff within the file related to Github, and the rest within gradle. It works with environment variables too, though, if you don't like the `-P` option.

PhilippSchmelter replied:
Should work like that now ;)

sebastian-peter commented on gradle/scripts/BranchName.gradle:
```diff
@@ -0,0 +1,23 @@
+      throw new GradleException("Error: Check Branch name format (e.g., ps/#1337-FeatureName).")
+    }
+
+    println "Branch name is $branchName"
+    project.ext.set('branch_name', branchName)
```
Is this still required or did you use this for testing?

PhilippSchmelter replied:
for testing again. I think it wouldn't hurt to let line 20 in. It's nice to just be able to read the captured branch name in the pipeline.

sebastian-peter reviewed (changes requested):
Something I just realized... The check will fail in dev, main and release branches. 

You can get a hint of what is needed in Jenkinsfile:
https://github.com/ie3-institute/simona/blob/741a2b2b5a7c54b7e59c1a9c71931f1eedb02e33/Jenkinsfile#L666-L671

PhilippSchmelter commented:
> Something I just realized... The check will fail in dev, main and release branches.
> 
> You can get a hint of what is needed in Jenkinsfile:
> 
> https://github.com/ie3-institute/simona/blob/741a2b2b5a7c54b7e59c1a9c71931f1eedb02e33/Jenkinsfile#L666-L671

I think the newest commit will implement the needed functionality

PhilippSchmelter commented:
Had some problems with spotless. Now it should really work for all kinds our branches.. 
This pull request introduces a new Gradle task to validate branch names and integrates this validation into the CI workflow. The key changes are:

### CI Workflow Enhancements:
* [`.github/workflows/ci.yml`](diffhunk://#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR33-L41): Added a step to check the branch name format during the CI process. This step determines the branch name based on the event type and then runs the new Gradle task to validate it.

### Gradle Task Addition:
* [`gradle/scripts/branchName.gradle`](diffhunk://#diff-69fa5c90d412548e5b537d4e7cedf6f1ed05368554c105b56664dcd46bcadd75R1-R26): Introduced a new Gradle task `checkBranchName` that validates the branch name against a set of predefined patterns. If the branch name does not match any pattern, an error is thrown.

sebastian-peter reviewed (approved):
Very nice, thanks!
